src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/haflow/sync/actions/UpdateHaFlowStatusAction.java

```diff
@@ -0,0 +1,151 @@
+import org.openkilda.messaging.command.haflow.HaFlowSyncResponse;
+import org.openkilda.messaging.command.haflow.HaFlowSyncResponse.HaFlowSyncResponseBuilder;
+import org.openkilda.messaging.info.InfoMessage;
+import org.openkilda.model.FlowPath;
+import org.openkilda.model.FlowPathStatus;
+import org.openkilda.model.HaFlow;
+import org.openkilda.model.HaFlowPath;
+import org.openkilda.model.PathId;
+import org.openkilda.model.SwitchId;
+import org.openkilda.persistence.PersistenceManager;
+import org.openkilda.wfm.share.logger.FlowOperationsDashboardLogger;
+import org.openkilda.wfm.topology.flowhs.fsm.common.actions.haflow.HaFlowProcessingWithHistorySupportAction;
+import org.openkilda.wfm.topology.flowhs.fsm.haflow.sync.HaFlowSyncContext;
+import org.openkilda.wfm.topology.flowhs.fsm.haflow.sync.HaFlowSyncFsm;
+import org.openkilda.wfm.topology.flowhs.fsm.haflow.sync.HaFlowSyncFsm.Event;
+import org.openkilda.wfm.topology.flowhs.fsm.haflow.sync.HaFlowSyncFsm.State;
+import org.openkilda.wfm.topology.flowhs.model.CrossingPaths;
+
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.UUID;
+
+@Slf4j
+public class UpdateHaFlowStatusAction extends
+        HaFlowProcessingWithHistorySupportAction<HaFlowSyncFsm, State, Event, HaFlowSyncContext> {
+    private final FlowOperationsDashboardLogger dashboardLogger;
+
+    public UpdateHaFlowStatusAction(
+            PersistenceManager persistenceManager, FlowOperationsDashboardLogger dashboardLogger) {
+        super(persistenceManager);
+        this.dashboardLogger = dashboardLogger;
+    }
+
+    @Override
+    protected void perform(
+            State from, State to, Event event, HaFlowSyncContext context, HaFlowSyncFsm stateMachine) {
+
+        Set<PathId> unsyncedPathIds = collectPathIdOfUnsuccessfulCommands(stateMachine);
+        HaFlow haflow = transactionManager.doInTransaction(() -> {
+            HaFlow haFlow = getHaFlow(stateMachine.getHaFlowId());
+
+            for (HaFlowPath haFlowPath : haFlow.getUsedPaths()) {
+                List<FlowPathStatus> subPathStatuses = new ArrayList<>();
+                for (FlowPath subPath : haFlowPath.getSubPaths()) {
+                    FlowPathStatus subPathStatus = computeSubPathStatus(unsyncedPathIds, subPath.getPathId());
+                    subPath.setStatus(subPathStatus);
+                    subPathStatuses.add(subPathStatus);
+
+                }
+                haFlowPath.setStatus(computeHaFlowStatus(unsyncedPathIds, haFlowPath.getHaPathId(), subPathStatuses));
+            }
+
+            haFlow.recalculateHaSubFlowStatuses();
+            haFlow.setStatus(haFlow.computeStatus());
+            return haFlow;
+        });
+
+        dashboardLogger.onHaFlowStatusUpdate(stateMachine.getHaFlowId(), haflow.getStatus());
+        stateMachine.saveActionToHistory(format("The ha-flow status was set to %s", haflow.getStatus()));
```

Comment:
It's how the history is saved for the simple flow.
Maybe it worth to add some convenient method if the old habits are strong or you want to save a couple of lines (but not in the stateMachine there are too many responsibilities already)